src/ConnectedMode/Migration/Wizard/MigrationWizardController.cs

```diff
@@ -0,0 +1,56 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.ComponentModel.Composition;
+
+namespace SonarLint.VisualStudio.ConnectedMode.Migration.Wizard
+{
+    /// <summary>
+    /// Starts a wizard which guides a user through the migration progress.
+    /// </summary>
+    public interface IMigrationWizardController
+    {
+        /// <summary>
+        /// Raised when the wizard is closed after successfully finishing the migration process.
+        /// </summary>
+        event EventHandler MigrationWizardFinished;
+
+        void StartMigrationWizard();
+    }
+
+    [Export(typeof(IMigrationPrompt))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class MigrationWizardController : IMigrationWizardController
```

Comment:
I got a bit confused at some point about the life cycle of this class. Do we want it to be a singleton? How long does it live? Always? Whenever a solution is closed / opened?

Reply:
Hhmmm. Good question. We only want one instance to be executing at a time, and we don't need to create it at all unless and until the user clicks on the `Migrate` button. We have two options:

* leave it as a singleton. If the user never clicks on `Migrate` we'll have created it unnecessarily, but that isn't a big problem. We'd also be relying on the fact that the migration uses a modal dialogue, which would prevent the user from triggering a second migration while the first one is active. If the user migrates several solutions one after another, they would all use the same instance.

* create it when needed i.e. create a new instance for each migration, and dispose of it when the migration finishes. That feels neater in some ways, but we'd need to make some changes to make it happen e.g. we'd need to create single factory to create the necessary migration objects when needed. Let's discuss it offline and work out what changes we would need to make.